sim/midas/src/main/scala/midas/passes/WriteXDCFile.scala

```diff
@@ -0,0 +1,102 @@
+// See LICENSE for license details.
+
+package midas.passes
+
+import firrtl._
+import firrtl.stage.Forms
+import firrtl.annotations._
+import firrtl.analyses.InstanceKeyGraph
+
+import midas.stage.{GoldenGateFileEmission}
+import midas.targetutils.xdc._
+
+/**
+  * We could reuse [[GoldenGateOutputFileAnnotation]] here, but this makes it
+  * marginally easier to filter out.
+  */
+private[midas] case class XDCOutputAnnotation(fileBody: String, suffix: Option[String])
+    extends NoTargetAnnotation with GoldenGateFileEmission {
+  def getBytes = fileBody.getBytes
+}
+
+private[midas] object WriteXDCFile extends Transform with DependencyAPIMigration with XDCAnnotationConstants {
+	override def prerequisites = Forms.LowForm
+  // Probably should run before emitter?
+
+  private def formatArguments(
+      iGraph: InstanceKeyGraph,
+      argumentList: Iterable[ReferenceTarget],
+      pathToCircuit: Option[String]): Iterable[Iterable[String]] = {
+
+    // Avoid the complexity of managing duplication of targets that don't
+    // explicitly agree on a root module.
+    //
+    // A better approach would be to find the LCA of all targets, and ensure none of them
+    // are duplicated under that ancestor. In practice, XDC snippets will be
+    // emitted with targets rooted at the current chisel module...
+    val rootModule = argumentList.head.module
+    // Explicit here means the LCA is encoded directly in the module field of the target.
+    val hasCommonExplicitRoot = argumentList.foldLeft(true) { _ && _.module == rootModule }
+    require(hasCommonExplicitRoot,
+      "All targets in an XDC Annotation must be rooted at the same module. Got:\n" +
+      argumentList.mkString("\n"))
+
+    // Get local paths for each reference under the enclosing module
+    val relativePaths = for (arg <- argumentList) yield {
+      arg.path.map { _._1.value } ++: // Path in the provided RT
+      arg.ref +:
+      arg.component.map { _.value }
+    }
+    // Prepare to duplicate the XDC snippet for each instance of the root module
+    val instances = iGraph.findInstancesInHierarchy(rootModule)
+    for (instPath <- instances) yield {
+      val fullInstPath = pathToCircuit ++: instPath.tail.map { _.Instance.value }
+      for (rPath <- relativePaths) yield {
+        (fullInstPath ++ rPath).mkString("/")
+      }
+    }
+  }
+
+  private def serializeXDC(
+      anno: XDCAnnotation,
+      iGraph: InstanceKeyGraph,
+      pathToCircuit: Option[String]): Iterable[String] = {
+    val segments = specifierRegex.split(anno.formatString)
+    val duplicatedArgumentLists = formatArguments(iGraph, anno.argumentList, pathToCircuit)
+    for (formattedArguments <- duplicatedArgumentLists) yield {
+      segments.zipAll(formattedArguments, "", "").foldLeft(""){ case (root, (a, b)) => root + a + b }
```

Comment:
```suggestion
      segments.zipAll(formattedArguments, "", "").map { case (a, b) => a + b }.mkString
```
A bit nitty but whatever.